src/dump.c

```diff
@@ -1015,41 +1015,35 @@ void jl_serialize_dependency_list(ios_t *s)
     static jl_array_t *deps = NULL;
     if (!deps)
         deps = (jl_array_t*)jl_get_global(jl_base_module, jl_symbol("_require_dependencies"));
-    if (deps) {
-        // sort!(deps) so that we can easily eliminate duplicates
-        static jl_value_t *sort_func = NULL;
-        if (!sort_func)
-            sort_func = jl_get_global(jl_base_module, jl_symbol("sort!"));
-        jl_apply((jl_function_t*)sort_func, (jl_value_t**)&deps, 1);
-
-        size_t l = jl_array_len(deps);
-        jl_value_t *prev = NULL;
+
+    // unique(deps) to eliminate duplicates while preserving order:
+    // we preserve order so that the topmost included .jl file comes first
+    static jl_value_t *unique_func = NULL;
+    if (!unique_func)
+        unique_func = jl_get_global(jl_base_module, jl_symbol("unique"));
+    jl_array_t *udeps = deps && unique_func ? (jl_array_t *) jl_apply((jl_function_t*)unique_func, (jl_value_t**)&deps, 1) : NULL;
```

Comment:
although I think it's safe here since we don't allocate any other memory before being done with this array, I'd be more comfortable if we rooted it anyway. That way it wont blow up if someone changes the code later to allocate something.